Update to Node 10 (10.4.0) and npm 6.1.0

blowery opened the pull request:
As an alternative to moving to just npm@6 in #25002, move up to the latest in the next LTS line, 10.4.0,. which included npm@6.1.0. 

~Holding off on adding the "needs e2e" label until we can get the basic test suite to pass.~ They passed! Added e2e tests. (@blowery)

## Testing Instructions
This is similar to most other node upgrades. See https://github.com/Automattic/wp-calypso/pull/25004, https://github.com/Automattic/wp-calypso/pull/23744, et all

### Local Dev
- Switch to node version 10.4.0 (npm 6.1.0) (an `nvm i` in the root will do it if you use `nvm`)
- `npm run distclean`
- `npm start`
- No functional changes to Calypso

### Local Docker Image
- Follow image instructions at: PCYsg-5XR-p2 or
- Setup docker locally and add the following host entry:
  ```
  127.0.0.1 wpcalypso.wordpress.com
  ```
- `npm run build-docker`
- `npm run docker`

### Desktop
- Clone the wp-desktop project
- Follow install instructions
- In the calypso submodule checkout this branch
- Desktop behaves normally

Desktop companion PR: https://github.com/Automattic/wp-desktop/pull/453

blowery committed:
Update to Node 10 (10.4.0) and npm 6.1.0

DanReyLop commented on package.json:
```diff
@@ -224,7 +224,6 @@
     "precommit": "npm run -s install-if-no-packages && node bin/pre-commit-hook.js",
     "prepush": "npm run -s install-if-no-packages && node bin/pre-push-hook.js",
     "rm": "node bin/rm.js",
-    "postshrinkwrap": "node bin/trim-shrinkwrap.js && touch node_modules",
```
`touch node_modules` was needed for `install-if-deps-outdated` to work. Please check if that's no longer needed.

blowery replied:
what's the expected behavior? the mtime on the node_modules directory should get updated after a `npm install` run?

blowery replied:
also, confirmed that `npm i --save some-package` updates npm-shinkwrap.json automatically.

DanReyLop replied:
> the mtime on the node_modules directory should get updated after a npm install run?

Yes, but after doing something like `npm run update-deps` or `npm install --save left-pad`, both `node_modules` and `npm-shrinkwrap.json` are updated, roughly at the same time. For `install-if-deps-outdated` to work, `node_modules` mtime must be newer or equal.

DanReyLop replied:
We are installing it already :)
https://github.com/Automattic/wp-calypso/pull/25374/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L253

DanReyLop replied:
It may be better to run a one-liner like `node -e 'require("fs").utimesSync("./node_modules", new Date(), new Date())'` than installing the `cash-touch` dependency just for this. What do you think?

blowery replied:
+1 on the one-liner in node

DanReyLop replied:
Done in https://github.com/Automattic/wp-calypso/pull/25374/commits/43d45904f675c878d1aa27a105ed16125739715d

blowery commented:
Hmm... in docker, this is having to build libsass from source which is slowing down the build quite a bit... curious to see if that gets cached off.

DanReyLop commented:
We need to update to the latest `node-sass` version, it has pre-built binaries for Node 10: https://github.com/sass/node-sass/releases/tag/v4.9.0

blowery committed:
pull in node-sass 4.9 which has node 10 support

blowery commented:
@DanReyLop quite right! done in https://github.com/Automattic/wp-calypso/pull/25374/commits/15b0549360c22ae73340f81f5e28dbb49d3f05da

sirreal commented:
It looks like `fsevents` also needs a bump for prebuild node 10 binaries: `fsevents@1.2.4`

blowery committed:
allow config as a builtin, even though it's not?

blowery committed:
update to fsevents 1.2.4 for node 10 support

blowery commented:
@sirreal done in https://github.com/Automattic/wp-calypso/pull/25374/commits/36746f2b7d9512ef021c8571638523cbbd144cb9 :)

blowery commented:
upgrading node-sass broke our css. Working on fixing that.

blowery committed:
fixup css that broke with node-sass 4.9

This also shows a bunch of deprecation warnings that look like they'll be removed in 4.9.1.

See https://github.com/sass/node-sass/issues/2362
Specifically https://github.com/sass/node-sass/issues/2362\#issuecomment-392118915

blowery commented:
fixed sass issues, though we now have lovely deprecation warnings, in https://github.com/Automattic/wp-calypso/pull/25374/commits/b47bb1873dfb13c7c55e0a59485caf13f46321a9

blowery commented on .eslintrc.js:
```diff
@@ -59,7 +59,7 @@ module.exports = {
 		// - events because we use it for some event emitters
 		// - path because we use it quite a bit
-		'import/no-nodejs-modules': [ 'error', { allow: [ 'url', 'events', 'path' ] } ],
+		'import/no-nodejs-modules': [ 'error', { allow: [ 'url', 'events', 'path', 'config' ] } ],
```
I'm not sure why this rule is complaining about config. It's not a builtin.

blowery committed:
touch node_modules after generating shrinkwrap

blowery commented:
Hmmm...

So the sass changes _compile_, but they don't _work_.

<img width="875" alt="screenshot 2018-06-08 09 55 47" src="https://user-images.githubusercontent.com/14350/41161865-2a6b5324-6b02-11e8-8d43-88cd2b1523bf.png">

Seeing some oddities with vars:
<img width="323" alt="screenshot 2018-06-08 09 56 53" src="https://user-images.githubusercontent.com/14350/41161916-49bee8a8-6b02-11e8-920b-eb2e378a2c53.png">

blowery commented:
TIL: We're using CSS Variables in Calypso. 

It appears that libsass is [no longer allowed to evaluate the value](https://github.com/sass/libsass/issues/2620) unless string interpolation is used.

Patch coming.

/cc @MichaelArestad  @frontdevde @nylen

MichaelArestad commented:
Whoa. What? That's a pain!

blowery committed:
use string interpolation on css variable values.

see https://github.com/sass/libsass/issues/2620

blowery commented:
@MichaelArestad right? though, have to say, i kinda dug the super stark calypso. :D

MichaelArestad commented:
@blowery you're not alone in that. Most of my other favorite apps have a similar feel.

frontdevde commented:
Thanks for catching / patching this, @blowery!
This also reminded me of color schemes in Calypso which I had almost forgotten about. A painful reminder that this feature will probably never see the light of day. 🙃🙈

Commit:
Removed "cash-touch" dependency, use instead a multi-platform NodeJS equivalent of "touch node_modules"
